deploy/main.go

```diff
@@ -0,0 +1,283 @@
+		if err != nil {
+			return nil, fmt.Errorf("failed to get app spec content: %w", err)
+		}
+		appSpecExpanded := os.ExpandEnv(string(appSpec))
+		if err := yaml.Unmarshal([]byte(appSpecExpanded), &spec); err != nil {
+			return nil, fmt.Errorf("failed to parse app spec: %w", err)
+		}
+	}
+
+	if err := replaceImagesInSpec(spec); err != nil {
+		return nil, fmt.Errorf("failed to replace images in spec: %w", err)
+	}
+	return spec, nil
+}
+
+// deploy deploys the app and waits for it to be live.
+func (d *deployer) deploy(ctx context.Context, spec *godo.AppSpec) (*godo.App, error) {
+	// Either create or update the app.
+	app, err := utils.FindAppByName(ctx, d.apps, spec.GetName())
+	if err != nil {
+		return nil, fmt.Errorf("failed to get app: %w", err)
+	}
+	if app == nil {
+		d.action.Infof("app %q does not exist yet, creating...", spec.Name)
+		app, _, err = d.apps.Create(ctx, &godo.AppCreateRequest{Spec: spec})
+		if err != nil {
+			return nil, fmt.Errorf("failed to create app: %w", err)
+		}
+	} else {
+		d.action.Infof("app %q already exists, updating...", spec.Name)
+		app, _, err = d.apps.Update(ctx, app.GetID(), &godo.AppUpdateRequest{Spec: spec})
+		if err != nil {
+			return nil, fmt.Errorf("failed to update app: %w", err)
+		}
+	}
+
+	ds, _, err := d.apps.ListDeployments(ctx, app.GetID(), &godo.ListOptions{PerPage: 1})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list deployments: %w", err)
+	}
+	if len(ds) == 0 {
+		return nil, fmt.Errorf("expected a deployment right after creating/updating the app, but got none")
+	}
+	// The latest deployment is the deployment we just created.
+	deploymentID := ds[0].GetID()
+
+	d.action.Infof("wait for deployment to finish")
+	dep, err := d.waitForDeploymentTerminal(ctx, app.ID, deploymentID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to wait deployment to finish: %w", err)
+	}
+
+	buildLogs, err := d.getLogs(ctx, app.ID, deploymentID, godo.AppLogTypeBuild)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get build logs: %w", err)
+	}
+	if len(buildLogs) > 0 {
+		d.action.SetOutput("build_logs", string(buildLogs))
+
+		if d.inputs.printBuildLogs {
+			d.action.Group("build logs")
+			d.action.Infof(string(buildLogs))
+			d.action.EndGroup()
+		}
+	}
+
+	deployLogs, err := d.getLogs(ctx, app.ID, deploymentID, godo.AppLogTypeDeploy)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get deploy logs: %w", err)
+	}
+	if len(deployLogs) > 0 {
+		d.action.SetOutput("deploy_logs", string(deployLogs))
+
+		if d.inputs.printDeployLogs {
+			d.action.Group("deploy logs")
+			d.action.Infof(string(deployLogs))
+			d.action.EndGroup()
+		}
+	}
+
+	if dep.Phase != godo.DeploymentPhase_Active {
+		// Fetch the app to get the latest state before returning.
+		app, _, err := d.apps.Get(ctx, app.ID)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get app after it failed: %w", err)
+		}
+		return app, fmt.Errorf("deployment failed: %s", dep.Phase)
```

Comment:
just to avoid any confusion
```suggestion
		return app, fmt.Errorf("deployment failed at phase: %s", dep.Phase)
```